L2/contract_upgrade_test.md

```diff
@@ -0,0 +1,201 @@
+
+1. **Familiarity with the upgrade workflow**, especially since we're using a customized `OptimismPortal2` contract, which may require additional steps beyond the standard flow.
+2. **Compatibility with Gnosis Safe**, ensuring the upgrade process works seamlessly when executed via a Gnosis Safe wallet.
+
+> **Important:** There is a one-to-one mapping between each OPCM and the target contracts version. You **must** deploy a dedicated OPCM for each version you plan to upgrade to. OPCM and `op-deployer` **do not support** upgrades between custom or non-standard versions.
+
+---
+
+## Step-by-Step Upgrade Procedure
+
+### 1. Create a Gnosis Safe Wallet (Sepolia)
+
+Create a Gnosis Safe on Sepolia testnet via https://app.safe.global and note the address. We will refer to it as:
+
+```bash
+$PROXY_ADMIN_OWNER
+```
+
+---
+
+### 2. Deploy `op-contracts/v2.0.0`
+
+```bash
+git clone https://github.com/ethereum-optimism/optimism.git optimism-v2
+cd optimism-v2
+git checkout op-contracts/v2.0.0
+
+mise install
+git submodule update --init --recursive
+
+cd op-deployer
+just build
+cd ..
+
+cd packages/contracts-bedrock
+forge clean
+just build
+cd ../..
+
+cd op-deployer
+./bin/op-deployer init --l1-chain-id 11155111 --l2-chain-ids $L2_CHAIN_ID --workdir .deployer --intent-config-type custom
+```
+
+#### Edit `.deployer/intent.toml`
+
+```toml
+configType = "custom"
+l1ChainID = 11155111
+fundDevAccounts = false
+useInterop = false
+l1ContractsLocator = "file:///root/optimism/packages/contracts-bedrock/forge-artifacts/"
+l2ContractsLocator = "file:///root/optimism/packages/contracts-bedrock/forge-artifacts/"
+
+[superchainRoles]
+proxyAdminOwner = "$PROXY_ADMIN_OWNER"
+protocolVersionsOwner = "$PROXY_ADMIN_OWNER"
+guardian = "$PROXY_ADMIN_OWNER"
+
+[[chains]]
+id = "0x000000000000000000000000000000000000000000000000000000000153c16e"
+baseFeeVaultRecipient = "$VAULT_RECIPIENT"
+l1FeeVaultRecipient = "$VAULT_RECIPIENT"
+sequencerFeeVaultRecipient = "$VAULT_RECIPIENT"
+eip1559DenominatorCanyon = 250
+eip1559Denominator = 50
+eip1559Elasticity = 6
+
+[chains.roles]
+l1ProxyAdminOwner = "$PROXY_ADMIN_OWNER"
+l2ProxyAdminOwner = "$PROXY_ADMIN_OWNER"
+systemConfigOwner = "$PROXY_ADMIN_OWNER"
+unsafeBlockSigner = "$SEQUENCER"
+batcher = "$BATCHER"
+proposer = "$PROPOSER"
+challenger = "$CHALLENGER"
+```
+
+#### Apply deployment
+
+```bash
+./bin/op-deployer apply   --workdir .deployer   --l1-rpc-url $L1_RPC_URL   --private-key $DEPLOYER_PRIVATE_KEY   --deployment-target live
+```
+
+Check deployed contracts in `.deployer/state.json`.
+
+---
+
+### 3. Prepare `op-contracts/v3.0.0` Artifacts
+
+Clone into a **new directory** to preserve the v2 deployment:
+
+```bash
+git clone https://github.com/ethereum-optimism/optimism.git optimism-v3
+cd optimism-v3
+git checkout op-contracts/v3.0.0
```

Comment:
For actual deployment, we'll need to merge our contract changes into the target release.